File: src/policy.ts
import fs from "node:fs";
import path from "node:path";

import yaml from "js-yaml";
import { parseArgsStringToArgv } from "string-argv";
import { z } from "zod";

import { PolicyError } from "./errors";

const pathsConfigSchema = z.object({
  projects_root: z
    .string()
    .min(1)
    .refine((value) => value.startsWith("/"), "path must be absolute"),
  data_root: z
    .string()
    .min(1)
    .refine((value) => value.startsWith("/"), "path must be absolute"),
  allowed_write_roots: z
    .array(
      z
        .string()
        .min(1)
        .refine((value) => value.startsWith("/"), "allowed_write_roots entries must be absolute paths"),
    )
    .default([]),
});

const agentConfigSchema = z.object({
  linux_user: z.string().min(1),
  allow_sudo: z.boolean().default(false),
  commands_whitelist: z.array(z.string().min(1)).default([]),
  sudo_commands_allowed: z.array(z.string().min(1)).default([]),
});

const policyConfigSchema = z.object({
  version: z.number().int(),
  paths: pathsConfigSchema,
  repos_allowed: z.array(z.string()).default([]),
  agents: z.record(agentConfigSchema),
  stacks_allowed: z.array(z.string()).default([]),
});

export type PolicyConfig = z.infer<typeof policyConfigSchema>;
export type AgentConfig = z.infer<typeof agentConfigSchema>;

export function loadPolicy(policyPath: string): PolicyConfig {
  if (!fs.existsSync(policyPath)) {
    throw new PolicyError(`policy file not found: ${policyPath}`);
  }

  let parsedRaw: unknown;
  try {
    parsedRaw = yaml.load(fs.readFileSync(policyPath, "utf-8"));
  } catch (error) {
    throw new PolicyError(`invalid YAML policy (${policyPath}): ${String(error)}`);
  }

  const parsed = policyConfigSchema.safeParse(parsedRaw);
  if (!parsed.success) {
    throw new PolicyError(`invalid policy schema (${policyPath}): ${parsed.error.message}`);
  }

  if (Object.keys(parsed.data.agents).length === 0) {
    throw new PolicyError("policy.agents cannot be empty");
  }

  return parsed.data;
}

function normalizeNameList(items: string[]): string[] {
  return items.map((item) => item.trim()).filter(Boolean);
}

export function normalizeCommand(command: string): { normalized: string; tokens: string[] } {
  if (!command || !command.trim()) {
    throw new PolicyError("command must be a non-empty string");
  }

  let tokens: string[];
  try {
    tokens = parseArgsStringToArgv(command);
  } catch (error) {
    throw new PolicyError(`invalid command syntax: ${String(error)}`);
  }

  if (tokens.length === 0) {
    throw new PolicyError("empty command after normalization");
  }

  return {
    normalized: tokens.join(" "),
    tokens,
  };
}

function normalizeAllowedCommands(commands: string[], context: string): Map<string, string[]> {
  const map = new Map<string, string[]>();
  for (const command of commands) {
    const { normalized, tokens } = normalizeCommand(command);
    map.set(normalized, tokens);
  }
  if (map.size === 0) {
    throw new PolicyError(`${context} is empty: DENY ALL`);
  }
  return map;
}

export function validateRepoAllowed(policy: PolicyConfig, repoName: string): string {
  const repo = (repoName ?? "").trim();
  if (!repo) {
    throw new PolicyError("repo name must be a non-empty string");
  }

  const reposAllowed = normalizeNameList(policy.repos_allowed);
  if (reposAllowed.length === 0) {
    throw new PolicyError("repos_allowed is empty: DENY ALL");
  }
  if (!reposAllowed.includes(repo)) {
    throw new PolicyError(`repo not allowed by policy: ${repo}`);
  }

  return repo;
}

export function validateAgent(policy: PolicyConfig, agentName: string): AgentConfig {
  const name = (agentName ?? "").trim();
  if (!name) {
    throw new PolicyError("agent name must be a non-empty string");
  }

  const agent = policy.agents[name];
  if (!agent) {
    throw new PolicyError(`agent not declared in policy: ${name}`);
  }

  return agent;
}

function isPathInside(basePath: string, targetPath: string, allowEqual: boolean): boolean {
  const relative = path.relative(basePath, targetPath);
  if (!relative) {
    return allowEqual;
  }
  return !relative.startsWith("..") && !path.isAbsolute(relative);
}

export function resolveRepoPath(policy: PolicyConfig, repoName: string): string {
  const repo = validateRepoAllowed(policy, repoName);
  const projectsRoot = path.resolve(policy.paths.projects_root);
  const resolved = path.resolve(projectsRoot, repo);

  if (!isPathInside(projectsRoot, resolved, false)) {
    throw new PolicyError(`resolved repo path escapes projects_root: ${resolved}`);
  }

  return resolved;
}

export function validatePathsWrite(policy: PolicyConfig, targetPath: string): string {
  const target = path.resolve(targetPath);
  const roots = normalizeNameList(policy.paths.allowed_write_roots).map((root) => path.resolve(root));

  if (roots.length === 0) {
    throw new PolicyError("allowed_write_roots is empty: DENY ALL writes");
  }

  for (const root of roots) {
    if (isPathInside(root, target, true)) {
      return target;
    }
  }

  throw new PolicyError(`path not allowed for write: ${target}`);
}

export function validateCommandWhitelist(
  policy: PolicyConfig,
  agentName: string,
  command: string,
): string[] {
  const agent = validateAgent(policy, agentName);
  const { normalized, tokens } = normalizeCommand(command);
  const allowed = normalizeAllowedCommands(
    agent.commands_whitelist,
    `commands_whitelist for ${agentName}`,
  );

  if (!allowed.has(normalized)) {
    throw new PolicyError(`command not in exact whitelist for ${agentName}: '${normalized}'`);
  }

  return tokens;
}

export function validateSudoCommand(policy: PolicyConfig, agentName: string, command: string): string {
  const agent = validateAgent(policy, agentName);
  if (!agent.allow_sudo) {
    throw new PolicyError(`agent ${agentName} is not allowed to use sudo`);
  }

  const { normalized } = normalizeCommand(command);
  const allowed = normalizeAllowedCommands(
    agent.sudo_commands_allowed,
    `sudo_commands_allowed for ${agentName}`,
  );
  if (!allowed.has(normalized)) {
    throw new PolicyError(`sudo command not in exact whitelist for ${agentName}: '${normalized}'`);
  }

  return normalized;
}

export function validateStack(policy: PolicyConfig, stackName: string): string {
  const stack = (stackName ?? "").trim();
  if (!stack) {
    throw new PolicyError("stack name must be a non-empty string");
  }

  const stacksAllowed = normalizeNameList(policy.stacks_allowed);
  if (stacksAllowed.length === 0) {
    throw new PolicyError("stacks_allowed is empty: DENY ALL");
  }

  if (!stacksAllowed.includes(stack)) {
    throw new PolicyError(`stack not allowed by policy: ${stack}`);
  }

  return stack;
}

File: shared/policy.py
#!/usr/bin/env python3
"""Strict policy loader and validators for ClawBrain."""

from __future__ import annotations

import shlex
from pathlib import Path
from typing import Dict, List, Tuple

import yaml
from pydantic import BaseModel, Field, ValidationError, field_validator


class PolicyError(Exception):
    """Raised when policy validation fails."""


class PathsConfig(BaseModel):
    projects_root: str
    data_root: str
    allowed_write_roots: List[str] = Field(default_factory=list)

    @field_validator("projects_root", "data_root")
    @classmethod
    def _validate_root_path(cls, value: str) -> str:
        if not isinstance(value, str) or not value.strip():
            raise ValueError("path must be a non-empty string")
        if not value.startswith("/"):
            raise ValueError("path must be absolute")
        return value.strip()

    @field_validator("allowed_write_roots")
    @classmethod
    def _validate_write_roots(cls, roots: List[str]) -> List[str]:
        normalized: List[str] = []
        for root in roots:
            if not isinstance(root, str) or not root.strip():
                raise ValueError("allowed_write_roots entries must be non-empty strings")
            entry = root.strip()
            if not entry.startswith("/"):
                raise ValueError("allowed_write_roots entries must be absolute paths")
            normalized.append(entry)
        return normalized


class AgentConfig(BaseModel):
    linux_user: str
    allow_sudo: bool = False
    commands_whitelist: List[str] = Field(default_factory=list)
    sudo_commands_allowed: List[str] = Field(default_factory=list)

    @field_validator("linux_user")
    @classmethod
    def _validate_linux_user(cls, value: str) -> str:
        if not isinstance(value, str) or not value.strip():
            raise ValueError("linux_user must be a non-empty string")
        return value.strip()


class PolicyConfig(BaseModel):
    version: int
    paths: PathsConfig
    repos_allowed: List[str] = Field(default_factory=list)
    agents: Dict[str, AgentConfig]
    stacks_allowed: List[str] = Field(default_factory=list)

    @field_validator("repos_allowed", "stacks_allowed")
    @classmethod
    def _validate_name_list(cls, items: List[str]) -> List[str]:
        normalized: List[str] = []
        for item in items:
            if not isinstance(item, str):
                raise ValueError("list entries must be strings")
            value = item.strip()
            if not value:
                continue
            normalized.append(value)
        return normalized


def load_policy(policy_path: str | Path) -> PolicyConfig:
    path = Path(policy_path)
    if not path.is_file():
        raise PolicyError(f"policy file not found: {path}")

    try:
        raw = yaml.safe_load(path.read_text(encoding="utf-8"))
    except yaml.YAMLError as exc:
        raise PolicyError(f"invalid YAML policy ({path}): {exc}") from exc
    except OSError as exc:
        raise PolicyError(f"failed reading policy ({path}): {exc}") from exc

    if not isinstance(raw, dict):
        raise PolicyError(f"policy must be a YAML mapping: {path}")

    try:
        policy = PolicyConfig.model_validate(raw)
    except ValidationError as exc:
        raise PolicyError(f"invalid policy schema ({path}): {exc}") from exc

    if not policy.agents:
        raise PolicyError("policy.agents cannot be empty")

    return policy


def normalize_command(command: str) -> Tuple[str, List[str]]:
    if not isinstance(command, str) or not command.strip():
        raise PolicyError("command must be a non-empty string")

    try:
        tokens = shlex.split(command, posix=True)
    except ValueError as exc:
        raise PolicyError(f"invalid command syntax: {exc}") from exc

    if not tokens:
        raise PolicyError("empty command after normalization")

    normalized = " ".join(tokens)
    return normalized, tokens


def validate_repo_allowed(policy: PolicyConfig, repo_name: str) -> str:
    repo = repo_name.strip() if isinstance(repo_name, str) else ""
    if not repo:
        raise PolicyError("repo name must be a non-empty string")

    if not policy.repos_allowed:
        raise PolicyError("repos_allowed is empty: DENY ALL")

    if repo not in policy.repos_allowed:
        raise PolicyError(f"repo not allowed by policy: {repo}")

    return repo


def validate_agent(policy: PolicyConfig, agent_name: str) -> AgentConfig:
    name = agent_name.strip() if isinstance(agent_name, str) else ""
    if not name:
        raise PolicyError("agent name must be a non-empty string")

    agent = policy.agents.get(name)
    if agent is None:
        raise PolicyError(f"agent not declared in policy: {name}")

    return agent


def _normalize_allowed_commands(commands: List[str], context: str) -> Dict[str, List[str]]:
    normalized: Dict[str, List[str]] = {}
    for command in commands:
        cmd_norm, cmd_tokens = normalize_command(command)
        normalized[cmd_norm] = cmd_tokens
    if not normalized:
        raise PolicyError(f"{context} is empty: DENY ALL")
    return normalized


def validate_command_whitelist(policy: PolicyConfig, agent_name: str, command: str) -> List[str]:
    agent = validate_agent(policy, agent_name)
    normalized_command, tokens = normalize_command(command)
    allowed = _normalize_allowed_commands(
        agent.commands_whitelist,
        f"commands_whitelist for {agent_name}",
    )

    if normalized_command not in allowed:
        raise PolicyError(
            f"command not in exact whitelist for {agent_name}: '{normalized_command}'"
        )

    return tokens


def validate_sudo_command(policy: PolicyConfig, agent_name: str, command: str) -> str:
    agent = validate_agent(policy, agent_name)
    if not agent.allow_sudo:
        raise PolicyError(f"agent {agent_name} is not allowed to use sudo")

    normalized_command, _ = normalize_command(command)
    allowed = _normalize_allowed_commands(
        agent.sudo_commands_allowed,
        f"sudo_commands_allowed for {agent_name}",
    )

    if normalized_command not in allowed:
        raise PolicyError(
            f"sudo command not in exact whitelist for {agent_name}: '{normalized_command}'"
        )

    return normalized_command


def resolve_repo_path(policy: PolicyConfig, repo_name: str) -> Path:
    repo = validate_repo_allowed(policy, repo_name)
    projects_root = Path(policy.paths.projects_root).resolve()
    resolved = (projects_root / repo).resolve()

    if resolved == projects_root or not resolved.is_relative_to(projects_root):
        raise PolicyError(f"resolved repo path escapes projects_root: {resolved}")

    return resolved


def validate_paths_write(policy: PolicyConfig, target_path: str | Path) -> Path:
    target = Path(target_path).resolve()
    allowed_roots = [Path(root).resolve() for root in policy.paths.allowed_write_roots]

    if not allowed_roots:
        raise PolicyError("allowed_write_roots is empty: DENY ALL writes")

    for root in allowed_roots:
        if target == root or target.is_relative_to(root):
            return target

    raise PolicyError(f"path not allowed for write: {target}")


def validate_stack(policy: PolicyConfig, stack_name: str) -> str:
    stack = stack_name.strip() if isinstance(stack_name, str) else ""
    if not stack:
        raise PolicyError("stack name must be a non-empty string")

    if not policy.stacks_allowed:
        raise PolicyError("stacks_allowed is empty: DENY ALL")

    if stack not in policy.stacks_allowed:
        raise PolicyError(f"stack not allowed by policy: {stack}")

    return stack

